File: agent/src/main/java/pt/um/lei/masb/agent/data/SoundCapturing.java
package pt.um.lei.masb.agent.data;

import javax.sound.sampled.*;
import java.math.BigDecimal;

public class SoundCapturing {

    private BigDecimal rms;
    private BigDecimal peak;

    public SoundCapturing() throws LineUnavailableException {
        short max = -1;
        int bytesRead;
        TargetDataLine line;
        int bufSize = 2048;

        //WAV format
        var format = new AudioFormat(AudioFormat.Encoding.PCM_SIGNED,
                                     44100,
                                     16,
                                     2,
                                     4,
                                     44100,
                                     false);
        var info = new DataLine.Info(TargetDataLine.class,
                                     format); // format is an AudioFormat object
        if (!AudioSystem.isLineSupported(info)) {
            System.out.println("Line not supported");
        }
        // Obtain and open the line.
        line = (TargetDataLine) AudioSystem.getLine(info);
        line.open(format, bufSize);
        line.start();
        byte[] buf = new byte[bufSize];
        float[] samples = new float[bufSize / 2];
        for (int b; (b = line.read(buf, 0, buf.length)) > -1; ) {

            // convert bytes to samples here
            for (int i = 0, s = 0; i < b; ) {
                int sample = 0;

                sample |= buf[i++] & 0xFF; // (reverse these two lines
                sample |= buf[i++] << 8;   //  if the format is big endian)

                // normalize to range of +/-1.0f
                samples[s++] = sample / 32768f;
            }
            double peak = 0;
            double rms = 0;
            for (float sample : samples) {

                double abs = Math.abs(sample);
                if (abs > peak) {
                    peak = abs;
                }

                rms += sample * sample;
            }

            this.rms = new BigDecimal(Math.sqrt(rms / samples.length));
            this.peak = new BigDecimal(peak);
        }
    }

    /**
     * @return Root mean square of recorded values, [-1, 1] range.
     */
    public BigDecimal getRms() {
        return rms;
    }

    /**
     * @return Peak of recorded values [0, 1].
     */
    public BigDecimal getPeak() {
        return peak;
    }
}

File: agent/src/main/kotlin/pt/um/masb/agent/data/SoundCapturing.kt
package pt.um.masb.agent.data

import mu.KotlinLogging
import java.math.BigDecimal
import javax.sound.sampled.AudioFormat
import javax.sound.sampled.AudioSystem
import javax.sound.sampled.DataLine
import javax.sound.sampled.TargetDataLine
import kotlin.experimental.and

private val logger = KotlinLogging.logger {}

fun captureSound(): Pair<BigDecimal, BigDecimal>? {
    var max: Short = -1
    var bytesRead: Int
    val line: TargetDataLine
    val bufSize = 2048

    //WAV format
    val format = AudioFormat(
        AudioFormat.Encoding.PCM_SIGNED,
        44100F,
        16,
        2,
        4,
        44100F,
        false
    )
    val info = DataLine.Info(TargetDataLine::class.java, format) // format is an AudioFormat object

    return if (!AudioSystem.isLineSupported(info)) {
        logger.error { "Line not supported" }
        null
    } else {
        // Obtain and open the line.
        line = AudioSystem.getLine(info) as TargetDataLine
        line.open(format, bufSize)
        line.start()

        val buf = ByteArray(bufSize)
        val samples = DoubleArray(bufSize / 2)
        val b = line.read(buf, 0, buf.size)
        while (b > -1) {

            // convert bytes to samples here
            var i = 0
            var s = 0
            while (i < b) {
                var sample = 0

                sample = sample or ((buf[i++] and 0xFF.toByte()).toInt()) // (reverse these two lines
                sample = sample or (buf[i++].toInt() shl 8)   //  if the format is big endian)

                // normalize to range of +/-1.0f
                samples[s++] = sample / 32768.0
            }
        }

        var peak = 0.0
        var rms = 0.0

        for (sample in samples) {

            val abs = Math.abs(sample)
            if (abs > peak) {
                peak = abs
            }

            rms += sample * sample
        }

        Pair(BigDecimal(Math.sqrt(rms / samples.size)), BigDecimal(peak))
    }
}